[jit] allow compilation using optional modules

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32539 [jit] allow compilation using optional modules**

Before: if something in `_modules` was `None`, we would barf. This is
incorrect because it's allowed for users to put `None` there, in case a
module is optional.

This case ought to be handled correctly during scripting. Fixes https://github.com/pytorch/pytorch/issues/32469

Differential Revision: [D19552346](https://our.internmc.facebook.com/intern/diff/D19552346)

suo committed:
[jit] allow compilation using optional modules

Before: if something in `_modules` was `None`, we would barf. This is
incorrect because it's allowed for users to put `None` there, in case a
module is optional.

This case ought to be handled correctly during scripting.

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 5758d394:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=5758d394d8ce2dfdac297e520f42de3337c1587e).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4324825) pytorch_xla_linux_xenial_py3_6_clang7_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=89899129) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=89899129))
<details>
<summary>
<code>Jan 23 18:32:16 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-device_test.diff</code>
</summary>

```
Jan 23 18:32:16 File to patch:  
Jan 23 18:32:16 Skip this patch? [y]  
Jan 23 18:32:16 4 out of 4 hunks ignored 
Jan 23 18:32:16 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 23 18:32:16 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-device_test.diff 
Jan 23 18:32:16 Traceback (most recent call last): 
Jan 23 18:32:16   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 67, in <module> 
Jan 23 18:32:16     patch_repo(args) 
Jan 23 18:32:16   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 49, in patch_repo 
Jan 23 18:32:16     raise RuntimeError('Failed to apply patch: {}'.format(ppath)) 
Jan 23 18:32:16 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-device_test.diff 
Jan 23 18:32:16 + cleanup 
Jan 23 18:32:16 + retcode=1 
Jan 23 18:32:16 + set +x 
Jan 23 18:32:16 =================== sccache compilation log =================== 
Jan 23 18:32:16 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 23 18:32:16 Compile requests                 1 
Jan 23 18:32:16 Compile requests executed        0 
Jan 23 18:32:16 Cache hits                       0 
Jan 23 18:32:16 Cache misses                     0 
Jan 23 18:32:16 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

eellison commented on test/jit/test_recursive_script.py:
```diff
@@ -650,3 +650,20 @@ def forward(self, input):
+        torch.jit.script(mod)
+
+        mod.foo = None
+        torch.jit.script(mod)
```
maybe trying running the module

eellison reviewed (approved):
LGTM

suo committed:
Update on "[jit] allow compilation using optional modules"


Before: if something in `_modules` was `None`, we would barf. This is
incorrect because it's allowed for users to put `None` there, in case a
module is optional.

This case ought to be handled correctly during scripting. Fixes https://github.com/pytorch/pytorch/issues/32469

[ghstack-poisoned]
